mitodl/mitxonline
# fix: Fix edX username validation to avoid username collision

#### Pre-Flight checklist


#### What are the relevant tickets?
Somewhat https://github.com/mitodl/mitxonline/issues/259 and this slack [discussion](https://mitodl.slack.com/archives/C02649X2P1V/p1674669465296569).
Fixes https://github.com/mitodl/mitxonline/issues/1393

#### What's this PR do?
Fixes the username validation with edX.

- We implemented username validation in https://github.com/mitodl/mitxonline/pull/757. We were rooting on a validation message.
- edX updated validation messages under https://github.com/openedx/edx-platform/pull/30937.
- This PR fixes the validation and also makes it generic so that any issue with the username can be validated.

#### How should this be manually tested?

- Make sure that your edX integration works and `FEATURES["ENABLE_UNICODE_USERNAME"] = True`
- Create an account in MITxOnline with accented characters.
- Verify that the account is created in MitxOnline and in edX.
- Now try to create another account with the same username but unaccented characters.
- You should see an error message that `A user already exists with this username. Please try a different one.`

#### Any background context you want to provide?
Most of the details are available in the slack [thread](https://mitodl.slack.com/archives/C02649X2P1V/p1674669465296569) and https://github.com/mitodl/mitxonline/issues/259

Here are the alternative approaches that I considered:

- Perform an unaccented search in MITxOnline in such cases by following https://docs.djangoproject.com/en/3.2/ref/contrib/postgres/lookups/#unaccent but that requires some setup in MitxOnline and Postgres.
- Perform an accented search in edX, but that might need some collation settings with MySQL that we are sure of.
